fastai/transforms.py

```diff
@@ -147,61 +153,139 @@ def __call__(self, x):
+        self.deg,self.mode,self.p = deg,mode,p
+        self.tfm_y=tfm_y
+
+    def set_state(self):
+        self.rdeg = rand0(self.deg)
+        self.rp = random.random()<self.p
+        print(self.rdeg, self.rp)
 
-    def do_transform(self, x, rot_times, do_flip):
-        x = np.rot90(x, rot_times)
-        return np.fliplr(x).copy() if do_flip else x
+    def do_transform(self, x):
+        if self.rp:
+            x = rotate_cv(x, self.rdeg, self.mode)
+        return x
 
+class RandomDihedralXY(CoordTransform):
+    def set_state(self):
+        self.rot_times = random.randint(0,3)
+        self.do_flip = random.random()<0.5
 
-def RandomFlip(): return lambda x: x if random.random()<0.5 else np.fliplr(x).copy()
+    def do_transform(self, x):
+        x = np.rot90(x, self.rot_times)
+        return np.fliplr(x).copy() if self.do_flip else x
 
-def channel_dim(x): return np.rollaxis(x, 2)
+class RandomFlipXY(CoordTransform):
+    def set_state(self):
+        self.do_flip = random.random()<0.5
+
+    def do_transform(self, x):
+        return np.fliplr(x).copy() if self.do_flip else x
+
+## TODO: talk to Jeremy about this
+class RandomLightingXY(CoordTransform):
+    def __init__(self, b, c, tfm_y=TfmType.NO):
+        self.tfm_y=tfm_y
+        self.b,self.c = b,c
+
+    def __call__(self, x, y):
+        self.set_state()
+        x,y = self.do_transform(x),y
+        return x, y 
+ 
+    def set_state(self):
+        self.b_rand = b = rand0(self.b)
+        self.c_rand = b = rand0(self.c)
+
+    def do_transform(self, x):
+        b = self.b
+        c = self.c
+        c = -1/(c-1) if c<0 else c+1
+        x = lighting(x, b, c)
+        return x
 
 def compose(im, fns):
     for fn in fns: im=fn(im)
     return im
 
+class CropType(IntEnum):
+    RANDOM = 1
+    CENTER = 2
+    NO = 3
+
 class Transforms():
-    def __init__(self, sz, tfms, denorm, rand_crop=False):
+    def __init__(self, sz, tfms, denorm, crop_type=CropType.CENTER):
         self.sz,self.denorm = sz,denorm
-        crop_fn = RandomCrop if rand_crop else CenterCrop
+        crop_fn = RandomCrop
+        if crop_type == CropType.CENTER: crop_fn = CenterCrop
+        if crop_type == CropType.NO: crop_fn = NoCrop
         self.tfms = tfms + [crop_fn(sz), channel_dim]
     def __call__(self, im, y): return compose(im, self.tfms), y
 
-def image_gen(normalizer, denorm, sz, tfms=None, max_zoom=None, pad=0):
+def image_gen(normalizer, denorm, sz, tfms=None, max_zoom=None, pad=0, crop_type=None):
     if tfms is None: tfms=[]
     elif not isinstance(tfms, collections.Iterable): tfms=[tfms]
     scale = [RandomScale(sz, max_zoom) if max_zoom is not None else Scale(sz)]
     if pad: scale.append(ReflectionPad(pad))
+    if max_zoom is not None and crop_type is not CropType.NO:
+        crop_type = CropType.RANDOM 
     return Transforms(sz+pad, scale + tfms + [normalizer], denorm,
-                      rand_crop=max_zoom is not None)
+                      crop_type)
```

Comment:
I see,
what about if I change the default of crop_type to CropType.CENTER. 

def image_gen(normalizer, denorm, sz, tfms=None, max_zoom=None, pad=0, crop_type=CropType.CENTER):

Then if is max_zoom is not None it will change it to crop_type = CropType.RANDOM when crop_type is not CropType.NO.
What do you think?